### PR TITLE
ramips: add support for D-Link COVR-X1860 A1

### PR DESCRIPTION
**Pull request by RolandoMagico**
The pull request is based on the discussion in https://forum.openwrt.org/t/add-support-for-d-link-covr-x1860

The COVR-X1860 are MT7621-based AX1800 devices (similar to DAP-X1860, but with two Ethernet ports and external power supply) that are sold in sets of two (COVR-X1862) and three (COVR-X1863).

Specification:
 - MT7621
 - MT7915 + MT7975 2x2 802.11ax (DBDC)
 - 256MB RAM
 - 128 MB flash
 - 3 LEDs (red, orange, white), routed to one indicator in the top of the device
 - 2 buttons (WPS in the back and Reset at the bottom of the device)

MAC addresses:
 - LAN MAC (printed on the device) is stored in config2 partition as ASCII (entry factory_mac=xx:xx:xx:xx:xx:xx)
 - WAN MAC: LAN MAC + 3
 - 2.4G MAC: LAN MAC + 1
 - 5G MAC: LAN MAC + 2

The pins for the serial console are already labeled on the board (VCC, TX, RX, GND). Serial settings: 3.3V, 115200,8n1

Flashing via OEM Web Interface:
 - Download openwrt-ramips-mt7621-dlink_covr-x1860-a1-squashfs-factory.bin via the OEM web interface firmware update
 - The configuration wizard can be skipped by directly going to http://192.168.0.1/UpdateFirmware_Simple.html

Flashing via Recovery Web Interface:
 - Set your IP address to 192.168.0.10, subnetmask 255.255.255.0
 - Press the reset button while powering on the deivce
 - Keep the reset button pressed until the status LED blinks red
 - Open a Chromium based browser and goto http://192.168.0.1
 - Download openwrt-ramips-mt7621-dlink_covr-x1860-a1-squashfs-recovery.bin

Revert back to stock using the Recovery Web Interface:
 - Set your IP address to 192.168.0.10, subnetmask 255.255.255.25
 - Press the reset button while powering on the deivce
 - Keep the reset button pressed until the status LED blinks red
 - Open a Chromium based browser and goto http://192.168.0.1
 - Flash a decrypted firmware image from D-Link. Decrypting an firmware image is described below.

Decrypting a D-Link firmware image:
 - Download https://github.com/openwrt/firmware-utils/blob/master/src/dlink-sge-image.c and https://raw.githubusercontent.com/openwrt/firmware-utils/master/src/dlink-sge-image.h
 - Compile a binary from the downloaded file, e.g. gcc dlink-sge-image.c -lcrypto -o dlink-sge-image
 - Run ./dlink-sge-image COVR-X1860 <OriginalFirmware> <OutputFile> -d
 - Example for firmware 102b01: ./dlink-sge-image COVR-X1860 COVR-X1860_RevA_Firmware_102b01.bin COVR-X1860_RevA_Firmware_102b01_Decrypted.bin -d

### CODE REVIEW

**Comment by aparcar** [comment]
Glad to see this moving so quickly now, let's wait for the CI and I'll compile this locally, too.

**Comment by aparcar** [comment]
@blocktrron could you please review this, too? I like that it is possible to flash from either recovery or  regular web interface but I don't remember that we usually offer those recovery images. What do you think? Also, would this be applicable for the DAP-1860?

**Comment by s-2** [comment]
Regarding MAC addresses, I guess I had not even correctly verified the WAN assignment on OEM firmware initially 😇 
(maybe confused it with M32, which has only two ports and automagically assigns LAN / WAN roles, thus just using one MAC).
So if you found this to be LAN MAC + 3, I believe it should be that way.

Regarding the commit message, please have a look at the characters per line limitation,
https://openwrt.org/submitting-patches#patch_checklist
it should be <75 for the description and even less for the title (which is often difficult and not always strictly enforced from what I could tell).

Maybe the forum link should go to the bottom of the commit message, so it starts directly with the device description.
It seems there is also a remainder of M32 in the decryption instructions 🙂 
`- Run ./m32-firmware-util M32`

Feel free to add my `Signed-Off-By` and/or `Co-Authored-By` (not sure which exactly to use when).

@aparcar For DAP-X1860, the factory image encryption is also known by now, but the GPL tarball was only published after the device was already merged, so it only supports recovery now. Someone could write a tool or build recipe for the encrypted factory image, I had added the encryption details in the wiki:
https://openwrt.org/inbox/toh/d-link/dap-x1860#factory_image_encryptiondecryption

**Comment by RolandoMagico** [comment]
@s-2 : Thanks for the hints, I updated the PR accordingly (hoping the line length doesn't cause issues - if they will, I'll fix that too).

**Comment by s-2** [comment]
The line length is more of an aesthetic thing, not really needed these days with larger screens, but I usually still try to keep the formatting concise, and with the 75-char limit.

One more thing I just remebered looking at my previous commit messages for other devices, we could probably add
```
skip wizard by directly accessing:
http://192.168.0.1/UpdateFirmware_Simple.html
```
to the install instructions here as well, though it should be confirmed on a real device running OEM once more.
Makes it a lot less painful to flash after unboxing a brand-new device, rather than setting wifi and new admin password etc. 🙂

**Comment by aparcar** [comment]
> http://192.168.0.1/UpdateFirmware_Simple.html

I think that's a nice advice to add to the commit message. Apart from that this looks fine to me.

**Comment by RolandoMagico** [comment]
> One more thing I just remebered looking at my previous commit messages for other devices, we could probably add
> 
> ```
> skip wizard by directly accessing:
> http://192.168.0.1/UpdateFirmware_Simple.html
> ```
> 
> to the install instructions here as well,

Done

**Comment by aparcar** [comment]
@blocktrron for me this looks fine, could you please give it a good look and merge if you think it's fine?

**Comment by raenye** [comment]
Typo in the filename in the following:
`Example for firmware 102b01: ./dlink-sge-image COVR-X1860 COVR-X1860_RevA_Firmware_102b01.bin COVR-X1860_RevA_Firmware_102b01_Derypted.bin -d`.



DAP-1620 rev B also supports encrypted firmware through regular interface and non encrypted firmware through web recovery, but there the web recovery image is different than sysupgrade image. Is it the same here?
If so, does it make sense to have 3 binaries (factory-regular, factory-recovery, and sysupgrade)?

(DAP-1720 rev A behaves similarly, but it's not yet merged)

**Comment by RolandoMagico** [comment]
> Typo in the filename in the following:

Thanks, fixed

File: target/linux/ramips/image/mt7621.mk
Comment (aparcar) [inline target/linux/ramips/image/mt7621.mk:653]: Is this needed since the device didn't exists before?

**Review by blocktrron — APPROVED** [review]

File: target/linux/ramips/dts/mt7621_dlink_covr-x1860-a1.dts
Comment (raenye) [inline target/linux/ramips/dts/mt7621_dlink_covr-x1860-a1.dts:26]: perhaps `button-reset`, etc.?

File: target/linux/ramips/dts/mt7621_dlink_covr-x1860-a1.dts
Comment (raenye) [inline target/linux/ramips/dts/mt7621_dlink_covr-x1860-a1.dts:42]: kinda confusing calling it `power` and `status` at the same time.
Perhaps `led-status_white`?

File: target/linux/ramips/dts/mt7621_dlink_covr-x1860-a1.dts
Comment (raenye) [inline target/linux/ramips/dts/mt7621_dlink_covr-x1860-a1.dts:43]: perhaps add `color` and `function` attributes and drop `label`?

File: target/linux/ramips/dts/mt7621_dlink_covr-x1860-a1.dts
Comment (raenye) [inline target/linux/ramips/dts/mt7621_dlink_covr-x1860-a1.dts:178]: Is this the official name of the port? not `wan`?

File: target/linux/ramips/dts/mt7621_dlink_covr-x1860-a1.dts
Comment (raenye) [inline target/linux/ramips/dts/mt7621_dlink_covr-x1860-a1.dts:183]: Is this the official name of the port? not `lan`?

File: target/linux/ramips/dts/mt7621_dlink_covr-x1860-a1.dts
Comment (s-2) [reply target/linux/ramips/dts/mt7621_dlink_covr-x1860-a1.dts:42]: indeed it should be `status_white` here to be consistent (although that node name in the dts is technically not used for anything, but it's still confusing).
It's often been a discussion whether to call an LED `power` or `status`, however it should in any case be consistent among LEDs that share the same physical position in the case (i.e. multi-colored)

File: target/linux/ramips/dts/mt7621_dlink_covr-x1860-a1.dts
Comment (s-2) [reply target/linux/ramips/dts/mt7621_dlink_covr-x1860-a1.dts:43]: is that how it's done upstream? I've never seen LEDs declared without label in OpenWrt

File: target/linux/ramips/dts/mt7621_dlink_covr-x1860-a1.dts
Comment (s-2) [reply target/linux/ramips/dts/mt7621_dlink_covr-x1860-a1.dts:178]: those are indeed the port labels printed on the case (curiously, the `internet` socket is yellow, since that's where the user is supposed to connect the yellow cable to their main router...)

Comment (raenye) [reply target/linux/ramips/dts/mt7621_dlink_covr-x1860-a1.dts:178]: `label` is deprecated: https://www.kernel.org/doc/Documentation/devicetree/bindings/leds/common.yaml

File: target/linux/ramips/dts/mt7621_dlink_covr-x1860-a1.dts
Comment (raenye) [reply target/linux/ramips/dts/mt7621_dlink_covr-x1860-a1.dts:43]: `label` is deprecated: https://www.kernel.org/doc/Documentation/devicetree/bindings/leds/common.yaml

File: target/linux/ramips/image/mt7621.mk
Comment (s-2) [reply target/linux/ramips/image/mt7621.mk:653]: apparently not, it was mostly a boilerplate found with the majority of the other devices (but indeed it just declares the COMPAT version, so could most probably be dropped)

File: target/linux/ramips/dts/mt7621_dlink_covr-x1860-a1.dts
Comment (raenye) [reply target/linux/ramips/dts/mt7621_dlink_covr-x1860-a1.dts:42]: So technically you could just do `led-1`, `led-2`, etc.
Maybe also reorder LEDs by GPIO?

File: target/linux/ramips/dts/mt7621_dlink_covr-x1860-a1.dts
Comment (raenye) [reply target/linux/ramips/dts/mt7621_dlink_covr-x1860-a1.dts:43]: But you should wait for #13837 before dropping the label property altogether.

File: target/linux/ramips/image/mt7621.mk
Comment (RolandoMagico) [reply target/linux/ramips/image/mt7621.mk]: I removed the line

File: target/linux/ramips/dts/mt7621_dlink_covr-x1860-a1.dts
Comment (RolandoMagico) [reply target/linux/ramips/dts/mt7621_dlink_covr-x1860-a1.dts:42]: How are dependencies to another MR handled usaullay? Should we wait until the other MR is done to apply the changes?

File: target/linux/ramips/dts/mt7621_dlink_covr-x1860-a1.dts
Comment (RolandoMagico) [reply target/linux/ramips/dts/mt7621_dlink_covr-x1860-a1.dts:43]: How are dependencies to another MR handled usaullay? Should we wait until the other MR is done to apply the changes?

File: target/linux/ramips/dts/mt7621_dlink_covr-x1860-a1.dts
Comment (RolandoMagico) [reply target/linux/ramips/dts/mt7621_dlink_covr-x1860-a1.dts:183]: yes, there are two ports on the device: Ethernet and Internet
image

**Comment by s-2** [comment]
finally found time to compile this again and upgrade my device from an old snapshot.
sysupgrade required force due to COMPAT_VERSION mismatch (caused by removal of `dsa-migration` include; but if I understand correctly, UCI configuration was never meant to be portable across devices in the first place? Otherwise, I would have expected all newer (DSA) devices to declare version 1.1 here and only swconfig devices remaining with 1.0).

Everything's working fine :slightly_smiling_face:

**Comment by aparcar** [comment]
Ideally someone like @adschm comments on this since he introduced it. I recall it's only important in case things are migrated

File: include/image-commands.mk
Comment (hauke) [inline include/image-commands.mk:231]: This was added in 4c0dc68f466588bb2ef60b29a4d8052b34112ca6, please rebase.

**Comment by hauke** [comment]
We have to define `dsa-migration` for all new mt7621 devices because we set the `ucidef_set_compat_version` to the 1.1, see https://github.com/openwrt/openwrt/blob/main/target/linux/ramips/mt7621/base-files/etc/board.d/05_compat-version#L11

File: include/image-commands.mk
Comment (RolandoMagico) [reply include/image-commands.mk]: Done

**Comment by RolandoMagico** [comment]
> We have to define `dsa-migration`

So it must be re-added?

**Comment by blocktrron** [comment]
@RolandoMagico yes, please. I will merge it afterwards.

**Comment by RolandoMagico** [comment]
@blocktrron : Done

File: target/linux/ramips/dts/mt7621_dlink_covr-x1860-a1.dts
Comment (hauke) [reply target/linux/ramips/dts/mt7621_dlink_covr-x1860-a1.dts:43]: You can create a new pull request changing this after  https://github.com/openwrt/openwrt/pull/13837 was merged.

**Comment by s-2** [comment]
Awesome :tada: 
I wonder how we should proceed in regard of a backport, the gluon user base would probably most appreciate it,
however I'm unsure about how to deal with the `firmware-utils` change here.

Would a patch be feasible in the backport, or in the downstream project (e.g. gluon)?
By the way, I wonder why older branches don't point to the latest `firmware-utils`, as the tools included there should be most backwards-compatible (in contrast to e.g. LuCI, which is always used in the current version through feeds, and sometimes does no longer build for older trees even within the same version).

**Pull request merged**
